### PR TITLE
Track purchase details on token mint

### PR DESCRIPTION
**Pull request by yoshiwarab**
## Description of the change
This PR adds a new PrimaryPurchaseDetails entity type that includes data about the minter address used to mint a token, as well as currency info. All core handlers have been updated to add this data. Pre-v3 contracts will not always have this populated as the necessary data is not always available from within the handler.

https://app.asana.com/0/1201568815538912/1205847400671374/f

>reminder: Any subgraph deployments should be documented as [Releases](https://github.com/ArtBlocks/artblocks-subgraph/releases) in this repository.

>reminder: Any changes to subgraph schema should be accompanied by corresponding changes to the Art Blocks documentation site, available at https://docs.artblocks.io/ (specifically, see the [Subgraph Entities](https://docs.artblocks.io/creator-docs/art-blocks-api/entities/) and [Subgraph Querying and API Overview](https://docs.artblocks.io/creator-docs/art-blocks-api/queries/) sections).

### CODE REVIEW

**Comment by yoshiwarab** [comment]
@ryley-o would you mind running the e2e tests when you get the chance, I've been unable to get the graph-node docker image to build on my new computer.

File: src/helpers.ts [hunk 1]
```diff
@@ -657,3 +659,85 @@ export function snapshotStateAtSettlementRevenueWithdrawal(
...
+      purchaseDetails.minterAddress = changetype<Bytes>(
+        ByteArray.fromHexString(minterConfig.minter)
+      );
+      purchaseDetails.currencyAddress = minterConfig.currencyAddress;
+      purchaseDetails.currencySymbol = minterConfig.currencySymbol;
+
+      purchaseDetails.save();
+      return purchaseDetails.id;
+    } else {
+      log.warning("Minter configuration with id {} does not exists", [
```
Comment (ryley-o) [inline src/helpers.ts:717]:
```suggestion
      log.warning("Minter configuration with id {} does not exist", [
```

File: schema.graphql [hunk 2]
```diff
@@ -349,6 +349,26 @@ type Token @entity {
 
   "Next available sale id"
   nextSaleId: BigInt!
+
+  primaryPurchaseDetails: PrimaryPurchaseDetails
+}
+
+type PrimaryPurchaseDetails @entity {
+  id: ID!
```
Comment (ryley-o) [inline schema.graphql:357]:
```suggestion
  "Unique identifier made up of contract address and token id"
  id: ID!
```

File: src/helpers.ts [hunk 3]
```diff
@@ -657,3 +659,85 @@ export function snapshotStateAtSettlementRevenueWithdrawal(
...
+      ]);
+      return null;
+    }
+  } else {
+    // If no minter configuration is associated with the project, use the minter currently assigned to the contract
+    const contract = Contract.load(project.contract);
+    const to = changetype<Address>(event.transaction.to || Address.zero());
+
+    // If the transaction to address is a whitelisted minter, use it as the minter address
+    if (contract && contract.mintWhitelisted.includes(to)) {
```
Comment (ryley-o) [inline src/helpers.ts:728]: In general, assuming that `tx.to` is the minter address is a good assumption, although it is not foolproof (e.g. bot minting from contract, etc.). I don't think it is worth adding much more complexity here, but perhaps before falling back to `Address.zero()`, you could check if the length of `contract.mintWhitelisted` == 1, in which case you could assume the single item in the array was the minter.

**Review by ryley-o — APPROVED** [review]
LGTM!

File: src/helpers.ts (hunk 3 above)
Comment (yoshiwarab) [reply src/helpers.ts]: > In general, assuming that tx.to is the minter address is a good assumption, although it is not bulletproof (e.g. bot minting from contract, etc.)

I think the check for `contract.mintWhitelisted.includes(to)` should confirm the assumption though no?

Also, added the check for only a single allowed minter, good call!

**Pull request merged**